gnovm/pkg/gnolang/debugger.go

```diff
@@ -0,0 +1,694 @@
+const listUsage = `list|l [locspec]`
+const listShort = `Show source code.`
+const listLong = `
+See 'help break' for locspec syntax. If locspec is empty,
+list shows the source code around the current line.
+`
+
+func debugList(m *Machine, arg string) (err error) {
+	loc := m.DebugLoc
+	hideCursor := false
+
+	if arg == "" {
+		debugLineInfo(m)
+		if m.lastDebugCmd == "up" || m.lastDebugCmd == "down" {
+			loc = debugFrameLoc(m, m.debugFrameLevel)
+			fmt.Fprintf(m.DebugOut, "Frame %d: %s:%d\n", m.debugFrameLevel, loc.File, loc.Line)
+		}
+	} else {
+		if loc, err = arg2loc(m, arg); err != nil {
+			return err
+		}
+		hideCursor = true
+		fmt.Fprintf(m.DebugOut, "Showing %s:%d\n", loc.File, loc.Line)
+	}
+	file, line := loc.File, loc.Line
+	lines, offset, err := sourceLines(file, line)
+	if err != nil {
+		return err
+	}
+	for i, l := range lines {
+		cursor := ""
+		if !hideCursor && file == loc.File && loc.Line == i+offset {
+			cursor = "=>"
+		}
+		fmt.Fprintf(m.DebugOut, "%2s %4d: %s\n", cursor, i+offset, l)
+	}
+	return nil
+}
+
+func debugLineInfo(m *Machine) {
+	line := string(m.Package.PkgName)
+	if len(m.Frames) > 0 {
+		f := m.Frames[len(m.Frames)-1]
+		if f.Func != nil {
+			line += "." + string(f.Func.Name) + "()"
+		}
+	}
+	fmt.Fprintf(m.DebugOut, "> %s %s:%d\n", line, m.DebugLoc.File, m.DebugLoc.Line)
+}
+
+const listLength = 10 // number of lines to display
+
+func sourceLines(name string, n int) ([]string, int, error) {
+	buf, err := os.ReadFile(name)
+	if err != nil {
+		return nil, 1, err
+	}
+	lines := strings.Split(string(buf), "\n")
+	start := max(1, n-listLength/2) - 1
+	end := min(start+listLength, len(lines))
+	return lines[start:end], start + 1, nil
+}
+
+func max(a, b int) int {
+	if a > b {
+		return a
+	}
+	return b
+}
+
+func min(a, b int) int {
+	if a < b {
+		return a
+	}
+	return b
+}
+
+// ---------------------------------------
+const printUsage = `print|p <expression>`
+const printShort = `Print a variable or expression.`
+
+func debugPrint(m *Machine, arg string) (err error) {
+	if arg == "" {
+		return errors.New("missing argument")
+	}
+	// Use the Go parser to get the AST representation of print argument as a Go expresssion.
+	ast, err := parser.ParseExpr(arg)
+	if err != nil {
+		return err
+	}
+	tv, err := debugEvalExpr(m, ast)
+	if err != nil {
+		return err
+	}
+	fmt.Println(m.DebugOut, tv)
+	return nil
+}
+
+// debugEvalExpr evaluates a Go expression in the context of the VM and returns
+// the corresponding value, or an error.
+// The supported expression syntax is a small subset of Go expressions:
+// basic literals, identifiers, selectors, index expressions, or a combination
+// of those are supported, but none of function calls, arithmetic, logic or
+// assign operations, type assertions of convertions.
+// This is sufficient for a debugger to perform 'print (*f).S[x][y]' for example.
+func debugEvalExpr(m *Machine, node ast.Node) (tv TypedValue, err error) {
```

Comment:
Using `Machine.Eval` here would cause a re-entrency problem, as we are already in step mode. Generating instructions by calling `Eval` for the purpose of parsing a debug command would cause at minimum un-necessary interferences. I do not see the benefit of using `ParseExpr` of `go2gno.go` in this particular and restricted context, where I do not expect the use of any specific Gno feature with regard to Go.

Reply:
First, thanks for noticing the issue of parsing literal rune and string, which need to be unquoted (now fixed).

As you suggested, I tried to reuse `parseExpr()` from `go2gno.go` and `Machine.Eval()` in something like:
```go
expr, err := ParseExpr(arg)
if err != nil {
    return err
}
m.Debugger.enabled = false  // avoid debugger re-entrency
defer func() { m.Debugger.enabled = true }()
res := m.Eval(expr)
fmt.Fprintln(m.Debugger.out, res[0])
```

It works in first approximation. However, there is at least 2 issues:

- If the expression syntax is valid, but semantic wrong (i.e printing an undeclared variable), then `parseExpr()` will pass but `Eval()` will panic. This is what we want in the interpreter for quick transaction abort, but not in the context of a debugger. Of course we could recover, but I'm worried of the possible leftovers, inconsistencies and corruptions that the debugger could leave in the VM, something that doesn't occur in the current code.

- The debugger can examine the content of data at different frame levels in the callstack, with `up` and `down` commands, using the `Debugger.framelevel` field to store the wanted level in a debugging session. This is broken by using `Eval()` which is not aware at all (and should not be IMO) of debugger metadata.

I think we could consider a code merge if and when the `Eval()` could return an error instead of panic, and if we are willing to let `Eval` work at different frame levels, and in restricted debugger mode (forbidding expressions which could alter machine state), but it starts to be intrusive and dangerous, for a tiny gain.

There is still some code deduplication opportunities, i.e `debugLookup()` as you noticed earlier, but I think it is wise to do that in a further step.